compiler/noirc_frontend/src/hir/comptime/value.rs

```diff
@@ -495,6 +495,11 @@ impl Value {
             Value::UnresolvedType(typ) => {
                 Token::InternedUnresolvedTypeData(interner.push_unresolved_type_data(typ))
             }
+            Value::TraitConstraint(trait_id, generics) => {
+                let name = Rc::new(interner.get_trait(trait_id).name.0.contents.clone());
+                let typ = Type::TraitAsType(trait_id, name, generics);
+                Token::QuotedType(interner.push_quoted_type(typ))
```

Comment:
Lowering this as a type made me a bit uncomfortable at first but I suppose it is fine since no data is lost. It does mean you could also use a `quote { impl Foo }.as_type()` where a trait constraint is expected which I suppose is fine.

Reply:
Ah, good point. I actually started doing it differently, having a new "interned" quote until I saw we have `Type::TraitAsType` which has `TraitId` and generics so I thought that was it. When printing it, it shows up as `impl Foo<..>` as you mentioned, so inside a quoted value in an `impl $constraint for ...` it looks odd, like "impl impl Foo for". So I think this concern is valid.

Would it make sense to introduce a different type here? Or we could add a boolean that says whether it's a constraint or not so we print it better but also disallow it for `quote { impl Foo }.as_type()`.

Reply:
I think keeping it the way it is would use the least code at least and is somewhat convenient for users. I don't think "impl Foo" being printed out is too confusing.